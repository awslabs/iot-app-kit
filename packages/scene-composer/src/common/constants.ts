import * as THREE from 'three';

import { Component, LightType } from '../models/SceneModels';
import { InfoIconSvgString, WarningIconSvgString, ErrorIconSvgString, VideoIconSvgString } from '../assets';
import {
  IValueDataBindingProviderState,
  DefaultAnchorStatus,
  DistanceUnit,
  Vector3,
  ITagSettings,
  IOverlaySettings,
} from '../interfaces';
import { CameraControlImpl } from '../store/internalInterfaces';

/******************************************************************************
 * Document Constants
 ******************************************************************************/

// TODO: [4553] remove in preview build
export const LEGACY_VERSION = '1';

export const CURRENT_MAJOR_VERSION = 1;
export const CURRENT_MINOR_VERSION = 0;
export const CURRENT_VERSION = `${CURRENT_MAJOR_VERSION}.${CURRENT_MINOR_VERSION}`;

// TODO: [4812] replace the url with the actual one
export const AWS_USER_GUIDE_DOC_URL = 'https://docs.aws.amazon.com/iot-twinmaker/latest/guide/what-is-twinmaker.html';

/******************************************************************************
 * DataBinding Constants
 ******************************************************************************/
export const DataBindingLabelKeys = {
  entityId: 'entityId',
  componentName: 'componentName',
  propertyName: 'propertyName',
  entityPath: 'entityPath',
  componentTypeId: 'componentTypeId',
};

export const EMPTY_VALUE_DATA_BINDING_PROVIDER_STATE: IValueDataBindingProviderState = {
  definitions: [],
  selectedOptions: [],
};

// For V1, the customer can only use 2 predefined data binding template in scene editor
export const DEFAULT_DATA_BINDING_TEMPLATE_ENTITY_ID = 'sel_entity';
export const DEFAULT_DATA_BINDING_TEMPLATE_COMPONENT_NAME = 'sel_comp';

/******************************************************************************
 * Scene Constants
 ******************************************************************************/
export const DEFAULT_DISTANCE_UNIT: DistanceUnit = 'meters';

export const ROOT_OBJECT_3D_NAME = 'ROOT';

export const DEFAULT_COLOR_MAPS = [
  new THREE.Color('red'),
  new THREE.Color('blue'),
  new THREE.Color('green'),
  new THREE.Color('crimson'),
  new THREE.Color('cyan'),
];

export const DEFAULT_CAMERA_SETTINGS = {
  fov: 53.13,
  near: 0.1,
  far: 1000.0,
  zoom: 1,
};

const DEFAULT_DIRECTIONAL_LIGHT_SETTINGS: Component.IDirectionalLightSettings = {
  color: 0xffffff,
  intensity: 1,
  castShadow: true,
};

const DEFAULT_POINT_LIGHT_SETTINGS: Component.IPointLightSettings = {
  color: 0xffffff,
  intensity: 1,
  decay: 2.0,
  distance: 0,
  castShadow: true,
};

const DEFAULT_AMBIENT_LIGHT_SETTINGS: Component.IAmbientLightSettings = {
  color: 0xffffff,
  intensity: 1,
};

const DEFAULT_HEMISPHERE_LIGHT_SETTINGS: Component.IHemisphereLightSettings = {
  color: 0xffffff,
  groundColor: 0x333333,
  intensity: 1,
};

export const DEFAULT_LIGHT_SETTINGS_MAP = {
  [LightType.Directional]: DEFAULT_DIRECTIONAL_LIGHT_SETTINGS,
  [LightType.Point]: DEFAULT_POINT_LIGHT_SETTINGS,
  [LightType.Hemisphere]: DEFAULT_HEMISPHERE_LIGHT_SETTINGS,
  [LightType.Ambient]: DEFAULT_AMBIENT_LIGHT_SETTINGS,
};

export const SCENE_ICONS: Record<DefaultAnchorStatus, string> = {
  [DefaultAnchorStatus.Info]: InfoIconSvgString,
  [DefaultAnchorStatus.Warning]: WarningIconSvgString,
  [DefaultAnchorStatus.Error]: ErrorIconSvgString,
  [DefaultAnchorStatus.Video]: VideoIconSvgString,
};

export enum RenderOrder {
  DrawEarly = -10,
  Default = 0,
  DrawLate = 10,
}

export enum Layers {
  RaycastAndRender = 0,
  RenderOnly = 1,
}

export const SCENE_BODY_CLASS = 'twinmaker_scene_container';

export const DRACO_PATH = 'https://www.gstatic.com/draco/versioned/decoders/1.4.1/';

export const DEFAULT_TAG_GLOBAL_SETTINGS: ITagSettings = {
  autoRescale: false,
  scale: 1,
};

export const DEFAULT_OVERLAY_GLOBAL_SETTINGS: IOverlaySettings = {
  overlayPanelVisible: false,
};

/******************************************************************************
 * Camera Constants
 ******************************************************************************/
export const DEFAULT_CAMERA_CONTROLS_OPTIONS: Pick<CameraControlImpl, 'dampingFactor' | 'maxDistance' | 'minDistance'> =
  {
    dampingFactor: 0.2,
    maxDistance: Infinity,
    minDistance: 0,
  };
export const DEFAULT_CAMERA_POSITION: Vector3 = [5, 5, 5];
export const DEFAULT_CAMERA_OPTIONS: Pick<THREE.PerspectiveCamera, 'far' | 'fov' | 'near'> = {
  far: 1000,
  fov: 53.13,
  near: 0.1,
};
export const DEFAULT_CAMERA_TARGET: Vector3 = [0, 0, 0];
export const DEFAULT_TWEEN_DURATION = 500;
export const SCENE_CAPABILITY_MATTERPORT = 'MATTERPORT';
export const SECRET_MANAGER_MATTERPORT_TAG = 'AWSIoTTwinMaker_Matterport';
export const MATTERPORT_SECRET_ARN = 'MATTERPORT_SECRET_ARN';
export const MATTERPORT_ACCESS_TOKEN = 'MATTERPORT_ACCESS_TOKEN';
export const MATTERPORT_APPLICATION_KEY = 'MATTERPORT_APPLICATION_KEY';
export const OPTIONS_PLACEHOLDER_VALUE = 'n/a';
