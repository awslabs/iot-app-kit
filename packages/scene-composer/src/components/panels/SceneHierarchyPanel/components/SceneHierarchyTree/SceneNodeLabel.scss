.tm-scene-node-label {
  .actions svg {
    cursor: pointer;
  }
}
