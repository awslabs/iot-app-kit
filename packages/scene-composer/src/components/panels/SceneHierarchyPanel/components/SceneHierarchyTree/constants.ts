import { KnownComponentType } from '../../../../../interfaces';
import draggable from '../../../../../enhancers/draggable';
import droppable from '../../../../../enhancers/droppable';
import Tree, { TreeItem } from '../../../../Tree';

export const AcceptableDropTypes = [
  ...Object.keys(KnownComponentType).filter((t) => t !== KnownComponentType.SubModelRef),
  'Space',
  'default',
];

export const EnhancedTreeItem = droppable(draggable(TreeItem));
export const EnhancedTree = droppable(Tree);
