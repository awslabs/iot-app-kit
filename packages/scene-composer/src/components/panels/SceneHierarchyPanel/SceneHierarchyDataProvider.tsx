import React, { FC, createContext, useContext, useCallback, useState, useEffect } from 'react';
import { DndProvider } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';
import { isEmpty } from 'lodash';
import { Euler, Vector3, Quaternion } from 'three';

import { useSceneComposerId } from '../../../common/sceneComposerIdContext';
import { findComponentByType, getFinalTransform, isEnvironmentNode, Transform } from '../../../utils/nodeUtils';
import { ISceneNodeInternal, useNodeErrorState, useSceneDocument, useStore } from '../../../store';
import useLifecycleLogging from '../../../logger/react-logger/hooks/useLifecycleLogging';
import { KnownComponentType } from '../../../interfaces';
import { RecursivePartial } from '../../../utils/typeUtils';

import ISceneHierarchyNode from './model/ISceneHierarchyNode';

type SelectionMode = 'single' | 'multi';

interface ISceneHierarchyContext {
  rootNodes: ISceneHierarchyNode[];
  searchTerms: string;
  selected?: string;
  pathFromSelectedToRoot?: string[];
  selectionMode: SelectionMode;
  getChildNodes(parentRef: string): Promise<ISceneHierarchyNode[]>;
  search(terms: string): void;
  select(objectRef: string): void;
  show(objectRef: string): void;
  hide(objectRef: string): void;
  activate(objectRef: string): void;
  unselect(objectRef: string): void;
  move(objectRef: string, newParentRef?: string);
  remove(objectRef: string);
  getObject3DBySceneNodeRef(objectRef: string);
  isViewing(): boolean;
  validationErrors: { [nodeRef: string]: string };
}

interface SceneHierarchyDataProviderProps {
  selectionMode: SelectionMode;
}

export const Context = createContext<ISceneHierarchyContext>({
  rootNodes: [],
  searchTerms: '',
  selectionMode: 'single',
  validationErrors: {},
  search: () => {},
  select: () => {},
  move: () => {},
  show: () => {},
  hide: () => {},
  activate: () => {},
  unselect: () => {},
  remove: () => {},
  getObject3DBySceneNodeRef: () => {},
  async getChildNodes() {
    return Promise.resolve([] as ISceneHierarchyNode[]);
  },
  isViewing: () => true,
});

export const useSceneHierarchyData = () => {
  return useContext(Context);
};

const toSceneHeirarchyNode = (
  { ref, name, parentRef, childRefs = [], components }: ISceneNodeInternal | Readonly<ISceneNodeInternal>,
  hideChild?: boolean,
) => {
  return {
    objectRef: ref,
    name,
    componentTypes: components.map((c) => c.type),
    childRefs: hideChild ? [] : childRefs,
    parentRef,
  } as ISceneHierarchyNode;
};

export const useChildNodes = (parentRef: string) => {
  const { getChildNodes } = useSceneHierarchyData();
  const [loading, setLoading] = useState(false);
  const [childNodes, setChildNodes] = useState([] as ISceneHierarchyNode[]);

  useEffect(() => {
    let mounted = true;
    (async () => {
      setLoading(true);
      const results = await getChildNodes(parentRef);
      if (mounted) {
        setChildNodes(results);
        setLoading(false);
      }
    })();
    return () => {
      mounted = false;
    };
  }, [getChildNodes]);

  return [childNodes, loading] as [ISceneHierarchyNode[], boolean];
};

const searchMatcher = (node: ISceneNodeInternal, terms: string) => {
  return node.name.toLowerCase().includes(terms.toLowerCase()); // Basic search matching algorithm;
};

const sortNodes = (a, b) => {
  const textA = a.name.toUpperCase();
  const textB = b.name.toUpperCase();
  return textA < textB ? -1 : textA > textB ? 1 : 0;
};

const SceneHierarchyDataProvider: FC<SceneHierarchyDataProviderProps> = ({ selectionMode, children }) => {
  const log = useLifecycleLogging('SceneHierarchyDataProvider');
  const sceneComposerId = useSceneComposerId();
  const { document, removeSceneNode } = useStore(sceneComposerId)((state) => state);
  const { isEditing } = useStore(sceneComposerId)((state) => state);
  const { updateSceneNodeInternal, updateDocumentInternal } = useSceneDocument(sceneComposerId);
  const selectedSceneNodeRef = useStore(sceneComposerId)((state) => state.selectedSceneNodeRef);
  const getSceneNodeByRef = useStore(sceneComposerId)((state) => state.getSceneNodeByRef);
  const getObject3DBySceneNodeRef = useStore(sceneComposerId)((state) => state.getObject3DBySceneNodeRef);
  const isViewing = useStore(sceneComposerId)((state) => state.isViewing);
  const { nodeMap } = document;
  const { nodeErrorMap: validationErrors } = useNodeErrorState(sceneComposerId);

  const rootNodeRefs = Object.values(nodeMap)
    .filter((item) => !item.parentRef && (!isEnvironmentNode(item) || isEditing()))
    .map((item) => item.ref);

  const [searchTerms, setSearchTerms] = useState('');
  const [filteredNodeMap, setFilteredNodeMap] = useState([] as ISceneNodeInternal[]);

  useEffect(() => {
    if (searchTerms === '') {
      setFilteredNodeMap([]);
    } else {
      const matchingNodes = Object.values(nodeMap).filter((node) => searchMatcher(node, searchTerms));
      setFilteredNodeMap(matchingNodes);
    }
  }, [nodeMap, searchTerms]);

  const [pathToRoot, setpathToRoot] = useState<string[]>([]);
  useEffect(() => {
    let currentNode = getSceneNodeByRef(selectedSceneNodeRef);
    const parentNodes: string[] = [];
    while (currentNode) {
      currentNode = getSceneNodeByRef(currentNode.parentRef);
      if (currentNode && parentNodes.indexOf(currentNode.ref) === -1) {
        parentNodes.push(currentNode.ref);
      }
    }
    setpathToRoot(parentNodes);
  }, [selectedSceneNodeRef]);

  const rootNodes: Readonly<ISceneNodeInternal>[] =
    filteredNodeMap.length > 0
      ? filteredNodeMap
      : rootNodeRefs
          .map(getSceneNodeByRef)
          .filter((node) => node !== undefined && searchMatcher(node, searchTerms))
          .map((item) => item as ISceneNodeInternal)
          .sort(sortNodes);

  const getChildNodes = useCallback(
    async (parentRef?: string) => {
      const results = Object.values(nodeMap)
        .filter((node) => node.parentRef === parentRef)
        .map((item) =>
          toSceneHeirarchyNode(item, Object.values(nodeMap).filter((n) => n.parentRef === item.ref).length > 0),
        )
        .sort(sortNodes);

      return Promise.resolve(results);
    },
    [getSceneNodeByRef, sceneComposerId, nodeMap, rootNodeRefs],
  );

  const activate = useCallback(
    (nodeRef: string) => {
      const setCameraTarget = useStore(sceneComposerId).getState().setCameraTarget;
      const node = getSceneNodeByRef(nodeRef);
      if (!isViewing() || !findComponentByType(node, KnownComponentType.Camera)) {
        setCameraTarget(nodeRef, 'transition');
      }
    },
    [sceneComposerId, isViewing],
  );

  const search = useCallback((terms: string) => {
    setSearchTerms(terms);
  }, []);

  const select = useCallback(
    (objectRef?: string) => {
      if (sceneComposerId) {
        useStore(sceneComposerId).getState().setSelectedSceneNodeRef(objectRef);
      }
    },
    [sceneComposerId],
  );

  const unselect = useCallback(() => {
    select(undefined); // TODO: Our existing state machine doesn't consider the possibility of multi-select
  }, []);

  const move = useCallback(
    (objectRef: string, newParentRef?: string) => {
      const originalObject = getSceneNodeByRef(objectRef);
      const originalObject3D = getObject3DBySceneNodeRef(objectRef);
      const oldParentRef = originalObject?.parentRef as string;

      if (oldParentRef === newParentRef) {
        log?.verbose('Parent ref unchanged. Do nothing.');
        return;
      }

      let newHierarchyParentObject = getObject3DBySceneNodeRef(newParentRef);
      const newParent = getSceneNodeByRef(newParentRef);

      // If the direct new parent is a sub model node, use its modelRef parent
      // to calculate the new transforms to prevent the object froming being
      // moved to a wrong position calculated from the sub model world space.
      if (findComponentByType(newParent, KnownComponentType.SubModelRef)) {
        let hierarchyParent = newParent;
        while (hierarchyParent && !findComponentByType(hierarchyParent, KnownComponentType.ModelRef)) {
          hierarchyParent = getSceneNodeByRef(hierarchyParent.parentRef);
        }
        newHierarchyParentObject = getObject3DBySceneNodeRef(hierarchyParent?.ref);
      }

      let maintainedTransform: Transform | null = null;
      if (originalObject3D) {
        const worldPosition = originalObject3D.getWorldPosition(new Vector3());
        const worldRotation = new Euler().setFromQuaternion(originalObject3D.getWorldQuaternion(new Quaternion()));
        const worldScale = originalObject3D.getWorldScale(new Vector3());
        maintainedTransform = getFinalTransform(
          {
            position: worldPosition,
            rotation: worldRotation,
            scale: worldScale,
          },
          newHierarchyParentObject,
        );
      }

      // Create updates to the moving object
      const partial: RecursivePartial<ISceneNodeInternal> = { parentRef: newParentRef };
      if (maintainedTransform) {
        // Scale of Tag component is independent of its ancestors, therefore keep its original value.
        const finalScale = findComponentByType(originalObject, KnownComponentType.Tag)
          ? originalObject?.transform.scale
          : maintainedTransform.scale.toArray();

        // Update the node position to remain in its world space
        partial.transform = {
          position: maintainedTransform.position.toArray(),
          rotation: maintainedTransform.rotation.toArray(),
          scale: finalScale,
        };
      }

      updateSceneNodeInternal(objectRef, partial);
    },
    [updateSceneNodeInternal, getSceneNodeByRef, nodeMap, document],
  );

  const show = useCallback(
    (objectRef: string) => {
      const object = getObject3DBySceneNodeRef(objectRef);
      if (object) {
        object.visible = true;
      }
    },
    [getObject3DBySceneNodeRef],
  );

  const hide = useCallback(
    (objectRef: string) => {
      const object = getObject3DBySceneNodeRef(objectRef);
      if (object) {
        object.visible = false;
      }
    },
    [getObject3DBySceneNodeRef],
  );

  const remove = useCallback(
    (objectRef: string) => {
      removeSceneNode(objectRef);
    },
    [removeSceneNode],
  );

  return (
    <DndProvider backend={HTML5Backend}>
      <Context.Provider
        value={{
          rootNodes: rootNodes.map((node) => toSceneHeirarchyNode(node, !isEmpty(searchTerms))),
          validationErrors,
          activate,
          selected: selectedSceneNodeRef,
          pathFromSelectedToRoot: pathToRoot,
          move,
          searchTerms,
          search,
          selectionMode,
          select,
          unselect,
          show,
          hide,
          remove,
          getChildNodes,
          getObject3DBySceneNodeRef,
          isViewing,
        }}
      >
        {children}
      </Context.Provider>
    </DndProvider>
  );
};

export default SceneHierarchyDataProvider;
