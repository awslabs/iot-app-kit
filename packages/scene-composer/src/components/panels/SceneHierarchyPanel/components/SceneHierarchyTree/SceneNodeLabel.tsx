import React, { FC, useCallback } from 'react';

import { KnownComponentType } from '../../../../../interfaces';
import useLogger from '../../../../../logger/react-logger/hooks/useLogger';
import { ReactComponent as ShowIcon } from '../../../../../assets/icons/show.svg';

import './SceneNodeLabel.scss';

interface SceneNodeLabelProps {
  labelText: string;
  componentTypes?: string[];
  visible?: boolean;
  onVisibilityChange?: (newVisibility: boolean) => void;
}

const SceneNodeLabel: FC<SceneNodeLabelProps> = ({
  labelText,
  componentTypes,
  visible,
  onVisibilityChange = () => {},
}) => {
  const log = useLogger('SceneNodeLabel');

  const toggleVisibility = useCallback(() => {
    onVisibilityChange(!visible);
  }, [visible, onVisibilityChange]);

  const componentTypeIcons = componentTypes
    ?.filter((type) => Object.keys(KnownComponentType).includes(type))
    .map(
      /* istanbul ignore next */ (type) => {
        try {
          // eslint-disable-next-line @typescript-eslint/no-var-requires
          const { ReactComponent: Icon } = require(`../../../../../assets/icons/${type.toLowerCase()}.svg`);
          return <Icon key={type} data-testid={`scene-node-type-${type}`} className='tm-icon-component-type' />;
        } catch (e) {
          // if we're missing an icon, or there's just some issue, swallow it, because it's really not that important.
          log?.error(`Failed to load "${type}" icon appears to be missing`, e);
          return null;
        }
      },
    );

  return (
    <span className={'tm-scene-node-label'}>
      {componentTypeIcons} {labelText}{' '}
      <span className='actions'>
        <ShowIcon className={visible ? 'visible' : 'hidden'} onClick={toggleVisibility} />
      </span>
    </span>
  );
};

SceneNodeLabel.displayName = 'SceneNodeLabel';

export default SceneNodeLabel;
