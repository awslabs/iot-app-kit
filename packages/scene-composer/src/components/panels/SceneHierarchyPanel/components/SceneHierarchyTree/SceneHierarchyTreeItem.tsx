import React, { FC, useCallback, useState } from 'react';

import ISceneHierarchyNode from '../../model/ISceneHierarchyNode';
import { useChildNodes, useSceneHierarchyData } from '../../SceneHierarchyDataProvider';
import { DropHandler } from '../../../../../hooks/useDropMonitor';

import SceneNodeLabel from './SceneNodeLabel';
import { AcceptableDropTypes, EnhancedTree, EnhancedTreeItem } from './constants';

interface SceneHierarchyTreeItemProps extends ISceneHierarchyNode {
  enableDragAndDrop?: boolean;
  expanded?: boolean;
}

const SceneHierarchyTreeItem: FC<SceneHierarchyTreeItemProps> = ({
  objectRef: key,
  name: labelText,
  componentTypes,
  enableDragAndDrop,
  expanded: defaultExpanded = true,
}: SceneHierarchyTreeItemProps) => {
  const [expanded, setExpanded] = useState(defaultExpanded);
  const [visible, setVisible] = useState(true);
  const [childNodes] = useChildNodes(key);

  const { selected, select, unselect, activate, move, show, hide, selectionMode } = useSceneHierarchyData();

  const onExpandNode = useCallback((expanded) => {
    setExpanded(expanded);
  }, []);

  const onToggle = useCallback(
    (newState: boolean) => {
      newState ? select(key) : unselect(key);
    },
    [selected, select, unselect, key],
  );

  const onActivated = useCallback(() => {
    activate(key);
    select(key);
  }, [key]);

  const dropHandler = useCallback<DropHandler<{ ref: string }>>(
    (item: { ref: string }, { beenHandled }) => {
      if (!beenHandled) {
        move(item.ref, key);
      }
    },
    [key],
  );

  const onVisibilityChange = useCallback(
    (newVisibility) => {
      if (newVisibility) {
        show(key);
      } else {
        hide(key);
      }

      setVisible(newVisibility);
    },
    [key, visible, show, hide],
  );

  return (
    <EnhancedTreeItem
      key={key}
      labelText={
        <SceneNodeLabel
          labelText={labelText}
          componentTypes={componentTypes}
          visible={visible}
          onVisibilityChange={onVisibilityChange}
        />
      }
      onExpand={onExpandNode}
      expanded={expanded}
      expandable={childNodes.length > 0}
      selected={selected === key}
      selectionMode={selectionMode}
      onSelected={onToggle}
      onActivated={onActivated}
      acceptDrop={AcceptableDropTypes}
      onDropped={dropHandler}
      draggable={enableDragAndDrop}
      dataType={componentTypes && componentTypes.length > 0 ? componentTypes[0] : /* istanbul ignore next */ 'default'} // TODO: This is somewhat based on the current assumption that items will currently only really have one componentType
      data={{ ref: key }}
    >
      {childNodes && expanded && (
        <EnhancedTree droppable={enableDragAndDrop} acceptDrop={AcceptableDropTypes} onDropped={dropHandler}>
          {childNodes.map((node) => (
            <SceneHierarchyTreeItem key={node.objectRef} enableDragAndDrop={enableDragAndDrop} {...node} />
          ))}
        </EnhancedTree>
      )}
    </EnhancedTreeItem>
  );
};

SceneHierarchyTreeItem.displayName = 'SceneNodeTreeItem';

export default SceneHierarchyTreeItem;
