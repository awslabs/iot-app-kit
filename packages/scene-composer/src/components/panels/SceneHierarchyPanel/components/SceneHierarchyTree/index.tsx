import React, { FC, useCallback } from 'react';

import { useSceneHierarchyData } from '../../SceneHierarchyDataProvider';
import { DropHandler } from '../../../../../hooks/useDropMonitor';

import { AcceptableDropTypes, EnhancedTree } from './constants';
import SceneHierarchyTreeItem from './SceneHierarchyTreeItem';

interface SceneHierarchyTreeProps {
  enableDragAndDrop: boolean;
}

const SceneHierarchyTree: FC<SceneHierarchyTreeProps> = ({ enableDragAndDrop }: SceneHierarchyTreeProps) => {
  const { rootNodes, move } = useSceneHierarchyData();

  const dropHandler = useCallback<DropHandler<{ ref: string }>>((item: { ref: string }, { beenHandled }) => {
    if (!beenHandled) {
      move(item.ref);
    }
  }, []);

  return (
    <EnhancedTree acceptDrop={AcceptableDropTypes} onDropped={dropHandler} droppable={enableDragAndDrop}>
      {rootNodes.map((root) => (
        <SceneHierarchyTreeItem key={root.objectRef} enableDragAndDrop={enableDragAndDrop} {...root} />
      ))}
    </EnhancedTree>
  );
};

SceneHierarchyTree.displayName = 'SceneHierarchyTree';

export default SceneHierarchyTree;
