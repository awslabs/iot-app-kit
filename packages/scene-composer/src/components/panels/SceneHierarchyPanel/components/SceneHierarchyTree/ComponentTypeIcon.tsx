import React from 'react';
import { Icon } from '@cloudscape-design/components';

import { Camera, Light, Modelref, Tag } from '../../../../../assets/auto-gen/icons';
import { KnownComponentType } from '../../../../../interfaces';

const ComponentTypeIcon = ({ type, ...props }: { type: string }) => {
  switch (type) {
    case KnownComponentType.Camera:
      return <Icon svg={<Camera {...props} />} />;
    case KnownComponentType.Light:
      return <Icon svg={<Light {...props} />} />;
    case KnownComponentType.ModelRef:
    case KnownComponentType.SubModelRef:
      return <Icon svg={<Modelref {...props} />} />;
    case KnownComponentType.Tag:
      return <Icon svg={<Tag {...props} />} />;
    default:
      return <></>;
  }
};

ComponentTypeIcon.displayName = ComponentTypeIcon;

export default ComponentTypeIcon;
