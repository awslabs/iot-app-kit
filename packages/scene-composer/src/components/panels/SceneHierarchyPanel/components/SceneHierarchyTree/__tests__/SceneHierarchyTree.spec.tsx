import React from 'react';
import { render } from '@testing-library/react';

import { useSceneHierarchyData } from '../../../SceneHierarchyDataProvider';
import { EnhancedTree } from '../constants';
import SceneHierarchyTree from '..';

jest.mock('../../../SceneHierarchyDataProvider');

// eslint-disable-next-line react/prop-types
jest.mock('../SceneHierarchyTreeItem', () => ({ children, ...props }) => (
  <div data-mocked='SceneHierarchyTreeItem' data-props={JSON.stringify(props)}>
    {children}
  </div>
));

jest.mock('../constants', () => ({
  EnhancedTree: jest.fn(),
}));

describe('<SceneHierarchyTree />', () => {
  it('should render as per snapshot', () => {
    (useSceneHierarchyData as any).mockImplementation(() => {
      return {
        rootNodes: [
          {
            objectRef: '1',
            someProp: 'yes',
          },
          {
            objectRef: '2',
            someProp: 'yes',
          },
          {
            objectRef: '3',
            someProp: 'yes',
          },
        ],
      };
    });

    (EnhancedTree as any).mockImplementation(({ children, ...props }) => (
      <div data-mocked='EnhancedTree' data-props={JSON.stringify(props)}>
        {children}
      </div>
    ));

    const { container } = render(<SceneHierarchyTree enableDragAndDrop={true} />);

    expect(container).toMatchSnapshot();
  });

  it('should reparent dropped items as rootNodes', () => {
    const moveFn = jest.fn();
    const droppedRef = '111';
    let dropHandler: Function = null as unknown as Function;

    (useSceneHierarchyData as any).mockImplementation(() => {
      return {
        rootNodes: [],
        move: moveFn,
      };
    });

    (EnhancedTree as any).mockImplementation(({ onDropped }) => {
      dropHandler = onDropped;

      return <div data-mocked='EnhancedTree' />;
    });

    render(<SceneHierarchyTree enableDragAndDrop={true} />);

    expect(dropHandler).not.toBeNull();

    if (dropHandler) {
      dropHandler({ ref: droppedRef }, { beenHandled: false });
    }

    expect(moveFn).toBeCalledWith(droppedRef);
  });
});
