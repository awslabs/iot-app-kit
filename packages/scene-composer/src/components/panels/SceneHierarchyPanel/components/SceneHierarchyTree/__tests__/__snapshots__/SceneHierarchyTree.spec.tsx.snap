// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SceneHierarchyTree /> should render as per snapshot 1`] = `
<div>
  <div
    data-mocked="EnhancedTree"
    data-props="{\\"droppable\\":true}"
  >
    <div
      data-mocked="SceneHierarchyTreeItem"
      data-props="{\\"enableDragAndDrop\\":true,\\"objectRef\\":\\"1\\",\\"someProp\\":\\"yes\\"}"
    />
    <div
      data-mocked="SceneHierarchyTreeItem"
      data-props="{\\"enableDragAndDrop\\":true,\\"objectRef\\":\\"2\\",\\"someProp\\":\\"yes\\"}"
    />
    <div
      data-mocked="SceneHierarchyTreeItem"
      data-props="{\\"enableDragAndDrop\\":true,\\"objectRef\\":\\"3\\",\\"someProp\\":\\"yes\\"}"
    />
  </div>
</div>
`;
